File: MATLAB/Tx/conv_enc.m
function [c] = conv_enc(bits,R)

N = length(bits);
A = zeros(N,1);
B = zeros(N,1);

T1 = 0;
T2 = 0;
T3 = 0;
T4 = 0;
T5 = 0;
T6 = 0;


c = zeros(2*N,1);
for n = 1:N
    % temporary storage of shift register states
    T1t = T1;
    T2t = T2;
    T3t = T3;
    T4t = T4;
    T5t = T5;
    A(n) = xor(bits(n),xor(xor(T2,T3),xor(T5,T6)));
    B(n) = xor(bits(n),xor(xor(T1,T2),xor(T3,T6)));
    
    % update shift registers
    T1 = bits(n);
    T2 = T1t;
    T3 = T2t;
    T4 = T3t;
    T5 = T4t;
    T6 = T5t;
    m = 1+(2*(n-1));
    c(m) = A(n);
    c(m+1) = B(n);
end

c = puncture_bits(c,R);
    

File: PYTHON/TX/conv_enc.py

import numpy as np
import cmath as cm


def conv_enc(bits, code_rate):
    
    N = len(bits)
    A = np.zeros(N)
    B = np.zeros(N)
    T1 = 0
    T2 = 0
    T3 = 0
    T4 = 0
    T5 = 0
    T6 = 0
    c = []
    for n  in range(N):
    # temporary storage of shift register states
        T1t = T1
        T2t = T2
        T3t = T3
        T4t = T4
        T5t = T5
        A[n] = int(bits[n])^ ((int(T2)^int(T3)) ^ (int(T5)^int(T6)))
        B[n]= int(bits[n])^ ((int(T1)^int(T2)) ^ (int(T3)^int(T6)))  # update shift registers
        T1 = bits[n];
        T2 = T1t
        T3 = T2t
        T4 = T3t
        T5 = T4t
        T6 = T5t


    if code_rate == '1/2':
        for n  in range(0,N):
            c.append(A[n])
            c.append(B[n])
           # c=[c,A[n],B[n]]
    elif code_rate=='2/3':
        for  n in range(0,N):
            if (n+1)%2 == 1:
                c.append(A[n])
                c.append(B[n])
            else:
               c.append( A[n])
        
    else:
        c.append( A[0])
        c.append(B[0])
        for n in range(1, N, 3):
            c.append( A[n])
            if (n+2) <= N:
                c.append( B[n+1])
                if (n+3) <= N:
                  c.append(A[n+2])
                  c.append(B[n+2])
    return  c
    

    
    
    
    
    
    
    
  


